extensions/binding/org.eclipse.smarthome.binding.bluetooth/src/main/java/org/eclipse/smarthome/binding/bluetooth/discovery/internal/BluetoothDiscoveryService.java

```diff
@@ -0,0 +1,178 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.binding.bluetooth.discovery.internal;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+import org.eclipse.smarthome.binding.bluetooth.BluetoothAdapter;
+import org.eclipse.smarthome.binding.bluetooth.BluetoothBindingConstants;
+import org.eclipse.smarthome.binding.bluetooth.BluetoothCompanyIdentifiers;
+import org.eclipse.smarthome.binding.bluetooth.BluetoothDevice;
+import org.eclipse.smarthome.binding.bluetooth.BluetoothDiscoveryListener;
+import org.eclipse.smarthome.binding.bluetooth.discovery.BluetoothDiscoveryParticipant;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.UID;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link BluetoothDiscoveryService} handles searching for BLE devices.
+ *
+ * @author Chris Jackson - Initial Contribution
+ * @author Kai Kreuzer - Introduced BluetoothAdapters and BluetoothDiscoveryParticipants
+ *
+ */
+@Component(immediate = true, service = DiscoveryService.class, configurationPid = "discovery.bluetooth")
+public class BluetoothDiscoveryService extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(BluetoothDiscoveryService.class);
+
+    private final static int SEARCH_TIME = 15;
+
+    private final Set<BluetoothAdapter> adapters = new CopyOnWriteArraySet<>();
+    private final Set<BluetoothDiscoveryParticipant> participants = new CopyOnWriteArraySet<>();
+    private final Map<UID, BluetoothDiscoveryListener> registeredListeners = new HashMap<>();
```

Comment:
Shouldn't this be a `ConcurrentHashMap`? Alternatively, the add... and remove... methods must be `synchronized`